/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT LOCAL MODULES/SUBWORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//
// MODULE: Local modules for DESeq2 normalization with multiple methods
//
include { NORMALIZE_DESEQ2_QC_INVARIANT_GENES } from '../../modules/local/normalize_deseq2_qc_invariant_genes'
include { NORMALIZE_DESEQ2_QC_ALL_GENES      } from '../../modules/local/normalize_deseq2_qc_all_genes'
include { NORMALIZE_DESEQ2_QC_INVARIANT_GENES as NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM } from '../../modules/local/normalize_deseq2_qc_invariant_genes'
include { NORMALIZE_DESEQ2_QC_ALL_GENES as NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM } from '../../modules/local/normalize_deseq2_qc_all_genes'
include { NORMALIZE_DESEQ2_QC_INVARIANT_GENES as NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME } from '../../modules/local/normalize_deseq2_qc_invariant_genes'
include { NORMALIZE_DESEQ2_QC_ALL_GENES as NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME } from '../../modules/local/normalize_deseq2_qc_all_genes'
include { NORMALIZE_DESEQ2_QC_INVARIANT_GENES as NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT } from '../../modules/local/normalize_deseq2_qc_invariant_genes'
include { NORMALIZE_DESEQ2_QC_ALL_GENES as NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT } from '../../modules/local/normalize_deseq2_qc_all_genes'
include { NORMALIZE_DESEQ2_QC_INVARIANT_GENES as NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO } from '../../modules/local/normalize_deseq2_qc_invariant_genes'
include { NORMALIZE_DESEQ2_QC_ALL_GENES as NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO } from '../../modules/local/normalize_deseq2_qc_all_genes'
include { DESEQ2_TRANSFORM                                  } from '../../modules/local/deseq2_transform'
include { DESEQ2_TRANSFORM as DESEQ2_TRANSFORM_STAR_GENOME  } from '../../modules/local/deseq2_transform'
include { DESEQ2_TRANSFORM as DESEQ2_TRANSFORM_HISAT2       } from '../../modules/local/deseq2_transform'
include { DESEQ2_TRANSFORM as DESEQ2_TRANSFORM_PSEUDO       } from '../../modules/local/deseq2_transform'
include { DESEQ2_SECTION_HEADER                                       } from '../../modules/local/deseq2_section_header'
include { DESEQ2_SECTION_HEADER as DESEQ2_SECTION_HEADER_STAR_GENOME } from '../../modules/local/deseq2_section_header'
include { DESEQ2_SECTION_HEADER as DESEQ2_SECTION_HEADER_HISAT2      } from '../../modules/local/deseq2_section_header'
include { DESEQ2_SECTION_HEADER as DESEQ2_SECTION_HEADER_PSEUDO      } from '../../modules/local/deseq2_section_header'
include { MULTIQC_CUSTOM_BIOTYPE             } from '../../modules/local/multiqc_custom_biotype'
include { GENOME_COUNT                       } from '../../modules/local/genome_count'
include { MERGE_GENOME_COUNTS                } from '../../modules/local/merge_genome_counts'
include { MERGE_GENOME_COUNTS as MERGE_GENOME_COUNTS_HISAT2 } from '../../modules/local/merge_genome_counts'
include { GENOME_COUNTS_REPORT               } from '../../modules/local/genome_counts_report'

// 
// MODULE: DeepTools BigWig normalization with DESeq2 scaling factors
//
include { DEEPTOOLS_BIGWIG_NORM              } from '../../modules/local/deeptools_bw_norm'
include { DEEPTOOLS_BIGWIG_NORM as DEEPTOOLS_BIGWIG_NORM_INVARIANT } from '../../modules/local/deeptools_bw_norm'
include { DEEPTOOLS_BIGWIG_NORM as DEEPTOOLS_BIGWIG_NORM_ALL_GENES } from '../../modules/local/deeptools_bw_norm'

//
// SUBWORKFLOW: Local subworkflows
//
include { ALIGN_STAR                            } from '../../subworkflows/local/align_star'
include { QUANTIFY_RSEM                         } from '../../subworkflows/local/quantify_rsem'
include { BAM_DEDUP_UMI as BAM_DEDUP_UMI_STAR   } from '../../subworkflows/nf-core/bam_dedup_umi'
include { BAM_DEDUP_UMI as BAM_DEDUP_UMI_HISAT2 } from '../../subworkflows/nf-core/bam_dedup_umi'

include { checkSamplesAfterGrouping      } from '../../subworkflows/local/utils_nfcore_rnaseq_pipeline'
include { multiqcTsvFromList             } from '../../subworkflows/nf-core/fastq_qc_trim_filter_setstrandedness'
include { biotypeInGtf                   } from '../../subworkflows/local/utils_nfcore_rnaseq_pipeline'
include { getInferexperimentStrandedness } from '../../subworkflows/local/utils_nfcore_rnaseq_pipeline'
include { methodsDescriptionText         } from '../../subworkflows/local/utils_nfcore_rnaseq_pipeline'
include { mapBamToPublishedPath          } from '../../subworkflows/local/utils_nfcore_rnaseq_pipeline'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    IMPORT NF-CORE MODULES/SUBWORKFLOWS
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

//
// MODULE: Installed modules from nf-core/modules
//
include { DUPRADAR                   } from '../../modules/nf-core/dupradar'
include { DUPRADAR as DUPRADAR_STAR  } from '../../modules/nf-core/dupradar'
include { DUPRADAR as DUPRADAR_HISAT2 } from '../../modules/nf-core/dupradar'
include { PRESEQ_LCEXTRAP            } from '../../modules/nf-core/preseq/lcextrap'
include { PRESEQ_LCEXTRAP as PRESEQ_LCEXTRAP_STAR } from '../../modules/nf-core/preseq/lcextrap'
include { PRESEQ_LCEXTRAP as PRESEQ_LCEXTRAP_HISAT2 } from '../../modules/nf-core/preseq/lcextrap'
include { QUALIMAP_RNASEQ            } from '../../modules/nf-core/qualimap/rnaseq'
include { QUALIMAP_RNASEQ as QUALIMAP_RNASEQ_STAR } from '../../modules/nf-core/qualimap/rnaseq'
include { QUALIMAP_RNASEQ as QUALIMAP_RNASEQ_HISAT2 } from '../../modules/nf-core/qualimap/rnaseq'
include { STRINGTIE_STRINGTIE        } from '../../modules/nf-core/stringtie/stringtie'
include { SUBREAD_FEATURECOUNTS as SUBREAD_FEATURECOUNTS_EXTRA } from '../../modules/nf-core/subread/featurecounts'
include { KRAKEN2_KRAKEN2 as KRAKEN2 } from '../../modules/nf-core/kraken2/kraken2/main'
include { BRACKEN_BRACKEN as BRACKEN } from '../../modules/nf-core/bracken/bracken/main'
include { MULTIQC                    } from '../../modules/nf-core/multiqc'
include { BEDTOOLS_GENOMECOV as BEDTOOLS_GENOMECOV_FW          } from '../../modules/nf-core/bedtools/genomecov'
include { BEDTOOLS_GENOMECOV as BEDTOOLS_GENOMECOV_REV         } from '../../modules/nf-core/bedtools/genomecov'
include { BEDTOOLS_GENOMECOV as BEDTOOLS_GENOMECOV_UNSTRANDED  } from '../../modules/nf-core/bedtools/genomecov'
include { SAMTOOLS_INDEX                                       } from '../../modules/nf-core/samtools/index'

//
// SUBWORKFLOW: Installed subworkflows from nf-core/modules
//
include { paramsSummaryMap                 } from 'plugin/nf-schema'
include { samplesheetToList                } from 'plugin/nf-schema'
include { paramsSummaryMultiqc             } from '../../subworkflows/nf-core/utils_nfcore_pipeline'
include { softwareVersionsToYAML           } from '../../subworkflows/nf-core/utils_nfcore_pipeline'
include { FASTQ_ALIGN_HISAT2               } from '../../subworkflows/nf-core/fastq_align_hisat2'
include { BAM_MARKDUPLICATES_PICARD        } from '../../subworkflows/nf-core/bam_markduplicates_picard'
include { BAM_MARKDUPLICATES_PICARD as BAM_MARKDUPLICATES_PICARD_STAR } from '../../subworkflows/nf-core/bam_markduplicates_picard'
include { BAM_MARKDUPLICATES_PICARD as BAM_MARKDUPLICATES_PICARD_HISAT2 } from '../../subworkflows/nf-core/bam_markduplicates_picard'
include { BAM_RSEQC                        } from '../../subworkflows/nf-core/bam_rseqc'
include { BAM_RSEQC as BAM_RSEQC_STAR      } from '../../subworkflows/nf-core/bam_rseqc'
include { BAM_RSEQC as BAM_RSEQC_HISAT2    } from '../../subworkflows/nf-core/bam_rseqc'
include { BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG as BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_FORWARD } from '../../subworkflows/nf-core/bedgraph_bedclip_bedgraphtobigwig'
include { BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG as BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_REVERSE } from '../../subworkflows/nf-core/bedgraph_bedclip_bedgraphtobigwig'
include { BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG as BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_UNSTRANDED } from '../../subworkflows/nf-core/bedgraph_bedclip_bedgraphtobigwig'
include { QUANTIFY_PSEUDO_ALIGNMENT                         } from '../../subworkflows/nf-core/quantify_pseudo_alignment'
include { FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS              } from '../../subworkflows/nf-core/fastq_qc_trim_filter_setstrandedness'

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    RUN MAIN WORKFLOW
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/

sample_status_header_multiqc    = file("$projectDir/workflows/rnaseq/assets/multiqc/sample_status_header.txt", checkIfExists: true)
ch_biotypes_header_multiqc      = file("$projectDir/workflows/rnaseq/assets/multiqc/biotypes_header.txt", checkIfExists: true)
ch_deseq2_pca_header            = file("$projectDir/workflows/rnaseq/assets/multiqc/deseq2_pca_header.txt", checkIfExists: true)
ch_deseq2_clustering_header     = file("$projectDir/workflows/rnaseq/assets/multiqc/deseq2_clustering_header.txt", checkIfExists: true)
ch_deseq2_read_dist_header      = file("$projectDir/workflows/rnaseq/assets/multiqc/read_distribution_normalized_header.txt", checkIfExists: true)
ch_dummy_file                   = sample_status_header_multiqc

workflow RNASEQ {

    take:
    ch_samplesheet       // channel: path(sample_sheet.csv)
    ch_versions          // channel: [ path(versions.yml) ]
    ch_fasta             // channel: path(genome.fasta)
    ch_gtf               // channel: path(genome.gtf)
    ch_fai               // channel: path(genome.fai)
    ch_chrom_sizes       // channel: path(genome.sizes)
    ch_gene_bed          // channel: path(gene.bed)
    ch_transcript_fasta  // channel: path(transcript.fasta)
    ch_star_index        // channel: path(star/index/)
    ch_rsem_index        // channel: path(rsem/index/)
    ch_hisat2_index      // channel: path(hisat2/index/)
    ch_kallisto_index    // channel: [ meta, path(kallisto/index/) ]
    ch_bbsplit_index     // channel: path(bbsplit/index/)
    ch_ribo_db           // channel: path(sortmerna_fasta_list)
    ch_sortmerna_index   // channel: path(sortmerna/index/)
    ch_splicesites       // channel: path(genome.splicesites.txt)
    ch_genomic_features  // channel: path(genomic_features.rds)
    ch_annotation_matrix // channel: path(annotation_matrix.txt)

    main:

    //
    //
    def valid_aligners = ['star', 'hisat2']
    def valid_pseudo_aligners = ['kallisto']
    def valid_quantifications = ['genome', 'rsem']
    def valid_normalization_methods = ['all_genes', 'invariant_genes', 'all_genes, invariant_genes']
    
    if (params.aligner && !valid_aligners.contains(params.aligner)) {
        error "Invalid aligner '${params.aligner}'. Valid options: ${valid_aligners.join(', ')}"
    }
    
    if (params.pseudo_aligner && !valid_pseudo_aligners.contains(params.pseudo_aligner)) {
        error "Invalid pseudo_aligner '${params.pseudo_aligner}'. Valid options: ${valid_pseudo_aligners.join(', ')}"
    }
    
    def quantification_methods = []
    if (params.quantification) {
        quantification_methods = params.quantification instanceof List ? 
            params.quantification : params.quantification.split(',').collect{it.trim()}
        
        def invalid_methods = quantification_methods.findAll { !valid_quantifications.contains(it) }
        if (invalid_methods.size() > 0) {
            error "Invalid quantification method(s): ${invalid_methods.join(', ')}. Valid options: ${valid_quantifications.join(', ')}"
        }
        
        quantification_methods = quantification_methods.unique()
        
        log.info "Selected quantification methods: ${quantification_methods.join(', ')}"
    } else {
        if (params.aligner && !params.skip_quantification_method) {
            quantification_methods = ['genome']
            log.info "No quantification methods specified, defaulting to: ${quantification_methods.join(', ')}"
        }
    }
    
    if (!valid_normalization_methods.contains(params.normalization_method)) {
        error "Invalid normalization_method '${params.normalization_method}'. Valid options: ${valid_normalization_methods.join(', ')}"
    }
    
    if (params.aligner == 'hisat2' && quantification_methods.size() > 0) {
        def incompatible_methods = quantification_methods.findAll { it != 'genome' }
        if (incompatible_methods.size() > 0) {
            error "HISAT2 aligner only supports 'genome' quantification. HISAT2 does not generate transcriptome BAM files required for RSEM or Salmon quantification. Incompatible methods: ${incompatible_methods.join(', ')}"
        }
    }
    
    if (!params.pseudo_aligner && !params.aligner) {
        error "Must specify at least one of: aligner (star/hisat2) or pseudo_aligner (kallisto)."
    }

    ch_multiqc_files = Channel.empty()
    ch_multiqc_star_files = Channel.empty()
    ch_multiqc_hisat2_files = Channel.empty()
    ch_multiqc_kallisto_files = Channel.empty()
    
    // Separate BAM channels per aligner for post-alignment QC
    ch_genome_bam_star = Channel.empty()
    ch_genome_bam_star_index = Channel.empty()
    ch_genome_bam_hisat2 = Channel.empty()
    ch_genome_bam_hisat2_index = Channel.empty()
    
    // Separate MultiQC channels per aligner
    ch_multiqc_shared = Channel.empty()          // Pre-alignment QC (FastQC, trimming, strandedness)
    ch_multiqc_star_qc = Channel.empty()         // STAR post-alignment QC
    ch_multiqc_star_quant = Channel.empty()      // STAR quantification
    ch_multiqc_hisat2_qc = Channel.empty()       // HISAT2 post-alignment QC
    ch_multiqc_hisat2_quant = Channel.empty()    // HISAT2 quantification
    
    ch_trim_status = Channel.empty()
    ch_map_status = Channel.empty()
    ch_strand_status = Channel.empty()
    ch_percent_mapped = Channel.empty()
    ch_scaling_factors = Channel.empty()
    ch_scaling_factors_individual = Channel.empty()
    ch_scaling_factors_individual_invariant = Channel.empty()
    ch_scaling_factors_individual_all_genes = Channel.empty()

    //
    //
    Channel
        .fromList(samplesheetToList(params.input, "${projectDir}/assets/schema_input.json"))
        .map {
            meta, fastq_1, fastq_2, genome_bam, transcriptome_bam ->
                if (!fastq_2) {
                    return [ meta.id, meta + [ single_end:true ], [ fastq_1 ], genome_bam, transcriptome_bam ]
                } else {
                    return [ meta.id, meta + [ single_end:false ], [ fastq_1, fastq_2 ], genome_bam, transcriptome_bam ]
                }
        }
        .groupTuple()
        .map { samplesheet ->
            checkSamplesAfterGrouping(samplesheet)
        }
        .branch {
            meta, reads, genome_bam, transcriptome_bam ->
                bam: params.skip_alignment && (genome_bam || transcriptome_bam)
                    return [ meta, genome_bam, transcriptome_bam ]
                fastq: reads.size() > 0 && reads[0]
                    return [ meta.findAll {it.key != 'percent_mapped'}, reads ]
        }
        .set { ch_input_branched }


    ch_fastq = ch_input_branched.fastq
    ch_genome_bam = ch_input_branched.bam.map { meta, genome_bam, transcriptome_bam -> [ meta, genome_bam ] }.distinct()
    ch_transcriptome_bam = ch_input_branched.bam.map { meta, genome_bam, transcriptome_bam -> [ meta, transcriptome_bam ] }.distinct()


    ch_percent_mapped = ch_input_branched.bam
        .filter{ meta, genome_bam, transcriptome_bam -> meta.percent_mapped }
        .map { meta, genome_bam, transcriptome_bam -> [ meta, meta.percent_mapped ] }

    SAMTOOLS_INDEX (
        ch_genome_bam
    )
    ch_genome_bam_index = params.bam_csi_index ? SAMTOOLS_INDEX.out.csi : SAMTOOLS_INDEX.out.bai
    ch_versions = ch_versions.mix(SAMTOOLS_INDEX.out.versions.first())

    //
    //

    FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS (
        ch_fastq,
        ch_fasta,
        ch_transcript_fasta,
        ch_gtf,
        Channel.empty(),  // ch_salmon_index - not used, salmon quantification removed
        ch_sortmerna_index,
        ch_bbsplit_index,
        ch_ribo_db,
        params.skip_bbsplit || ! params.fasta,
        params.skip_fastqc || params.skip_qc,
        params.skip_trimming,
        params.skip_umi_extract,
        true,
        false,
        params.trimmer,
        params.min_trimmed_reads,
        params.save_trimmed,
        params.remove_ribo_rna,
        params.with_umi,
        params.umi_discard_read,
        params.stranded_threshold,
        params.unstranded_threshold,
        params.skip_linting,
        false
    )

    // Route pre-alignment QC to shared channel (used by all aligner reports)
    ch_multiqc_shared                 = ch_multiqc_shared.mix(FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS.out.multiqc_files)
    ch_multiqc_files                  = ch_multiqc_files.mix(FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS.out.multiqc_files)
    ch_versions                       = ch_versions.mix(FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS.out.versions)
    ch_strand_inferred_filtered_fastq = FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS.out.reads
    ch_trim_read_count                = FASTQ_QC_TRIM_FILTER_SETSTRANDEDNESS.out.trim_read_count

    ch_trim_status = ch_trim_read_count
        .map {
            meta, num_reads ->
                return [ meta.id, num_reads > params.min_trimmed_reads.toFloat() ]
        }

    //
    //
    ch_star_log            = Channel.empty()
    ch_unaligned_sequences = Channel.empty()

    if (!params.skip_alignment && params.aligner == 'star') {
        def is_aws_igenome = false
        if (params.fasta && params.gtf) {
            if ((file(params.fasta).getName() - '.gz' == 'genome.fa') && (file(params.gtf).getName() - '.gz' == 'genes.gtf')) {
                is_aws_igenome = true
            }
        }

        ALIGN_STAR (
            ch_strand_inferred_filtered_fastq,
            ch_star_index.map { [ [:], it ] },
            ch_gtf.map { [ [:], it ] },
            params.star_ignore_sjdbgtf,
            '',
            params.seq_center ?: '',
            is_aws_igenome,
            ch_fasta.map { [ [:], it ] },
            params.use_sentieon_star
        )

        // Separate STAR BAM outputs for aligner-specific QC
        ch_genome_bam_star               = ch_genome_bam_star.mix(ALIGN_STAR.out.bam)
        ch_genome_bam_star_index         = ch_genome_bam_star_index.mix(params.bam_csi_index ? ALIGN_STAR.out.csi : ALIGN_STAR.out.bai)
        // Keep old channels for backward compatibility
        ch_genome_bam                    = ch_genome_bam.mix(ALIGN_STAR.out.bam)
        ch_genome_bam_index              = ch_genome_bam_index.mix(params.bam_csi_index ? ALIGN_STAR.out.csi : ALIGN_STAR.out.bai)
        ch_transcriptome_bam             = ch_transcriptome_bam.mix(ALIGN_STAR.out.bam_transcript)
        ch_percent_mapped                = ch_percent_mapped.mix(ALIGN_STAR.out.percent_mapped)
        ch_unprocessed_bams              = ch_genome_bam.join(ch_transcriptome_bam)
        ch_star_log                      = ALIGN_STAR.out.log_final
        ch_unaligned_sequences           = ALIGN_STAR.out.fastq
        ch_multiqc_star_files            = ch_multiqc_star_files.mix(ch_star_log.collect{it[1]})
        ch_multiqc_star_qc               = ch_multiqc_star_qc.mix(ch_star_log.collect{it[1]})

        ch_versions = ch_versions.mix(ALIGN_STAR.out.versions)

        //
        //
        if (params.with_umi) {

            BAM_DEDUP_UMI_STAR(
                ch_genome_bam.join(ch_genome_bam_index, by: [0]),
                ch_fasta.map { [ [:], it ] },
                params.umi_dedup_tool,
                params.umitools_dedup_stats,
                params.bam_csi_index,
                ch_transcriptome_bam,
                ch_transcript_fasta.map { [ [:], it ] }
            )

            ch_genome_bam        = BAM_DEDUP_UMI_STAR.out.bam
            ch_transcriptome_bam = BAM_DEDUP_UMI_STAR.out.transcriptome_bam
            ch_genome_bam_index  = params.bam_csi_index ? BAM_DEDUP_UMI_STAR.out.csi : BAM_DEDUP_UMI_STAR.out.bai
            ch_versions          = ch_versions.mix(BAM_DEDUP_UMI_STAR.out.versions)

            ch_multiqc_files = ch_multiqc_files
                .mix(BAM_DEDUP_UMI_STAR.out.multiqc_files)

        } else {

            ch_multiqc_files = ch_multiqc_files
                .mix(ALIGN_STAR.out.stats.collect{it[1]})
                .mix(ALIGN_STAR.out.flagstat.collect{it[1]})
                .mix(ALIGN_STAR.out.idxstats.collect{it[1]})
        }
    }

    //
    // SUBWORKFLOW: Quantify gene and transcript abundance with RSEM
    //
    if (params.aligner == 'star' && !params.skip_quantification_method && quantification_methods.size() > 0) {
        
        if (quantification_methods.contains('rsem')) {
            //
            // QUANTIFICATION: STAR alignment + RSEM quantification
            //
            QUANTIFY_RSEM (
                ch_transcriptome_bam,
                ch_rsem_index,
                params.use_sentieon_star,
                ch_annotation_matrix
            )
            ch_multiqc_files = ch_multiqc_files.mix(QUANTIFY_RSEM.out.stat.collect{it[1]})
            ch_multiqc_star_quant = ch_multiqc_star_quant.mix(QUANTIFY_RSEM.out.stat.collect{it[1]})
            ch_versions = ch_versions.mix(QUANTIFY_RSEM.out.versions)

            if (!params.skip_qc & !params.skip_deseq2_qc) {
                //
                // DESeq2 normalization: Parse normalization methods (all_genes, invariant_genes, or both)
                //
                def normalization_methods = params.normalization_method instanceof List ? 
                    params.normalization_method : params.normalization_method.split(',').collect{it.trim()}
                
                
                ch_normalization_versions = Channel.empty()
                ch_normalization_scaling_factors = Channel.empty()
                ch_deseq2_raw_files = Channel.empty()
                
                //
                // MODULE: Invariant genes normalization (stable genes only)
                //
                if (normalization_methods.contains('invariant_genes')) {
                    NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM (
                        QUANTIFY_RSEM.out.merged_counts_gene,
                        "STAR_RSEM"
                    )
                    
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.read_dist_norm_txt)
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.sample_distances_txt)
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.pca_all_genes_txt)
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.pca_top_genes_txt)
                    
                    ch_normalization_versions = ch_normalization_versions.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.versions)
                    ch_normalization_scaling_factors = ch_normalization_scaling_factors.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.scaling_factors)
                    ch_normalization_scaling_factors = ch_normalization_scaling_factors.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.scaling_factors_individual)
                    ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.scaling_factors_individual)
                    // Tag files with quantification method for downstream use
                    ch_scaling_factors_individual_invariant = ch_scaling_factors_individual_invariant.mix(
                        NORMALIZE_DESEQ2_QC_INVARIANT_GENES_RSEM.out.scaling_factors_individual
                            .flatten()
                            .map { file -> [file, 'rsem'] }
                    )
                }
                
                //
                // MODULE: All genes normalization (default DESeq2 method)
                //
                if (normalization_methods.contains('all_genes') || (!normalization_methods.contains('invariant_genes') && !normalization_methods.contains('all_genes'))) {
                    NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM (
                        QUANTIFY_RSEM.out.merged_counts_gene,
                        "STAR_RSEM"
                    )
                    
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.read_dist_norm_txt)
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.sample_distances_txt)
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.pca_all_genes_txt)
                    ch_deseq2_raw_files = ch_deseq2_raw_files.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.pca_top_genes_txt)
                    
                    ch_normalization_versions = ch_normalization_versions.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.versions)
                    ch_normalization_scaling_factors = ch_normalization_scaling_factors.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.scaling_factors)
                    ch_normalization_scaling_factors = ch_normalization_scaling_factors.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.scaling_factors_individual)
                    ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.scaling_factors_individual)
                    // Tag files with quantification method for downstream use
                    ch_scaling_factors_individual_all_genes = ch_scaling_factors_individual_all_genes.mix(
                        NORMALIZE_DESEQ2_QC_ALL_GENES_RSEM.out.scaling_factors_individual
                            .flatten()
                            .map { file -> [file, 'rsem'] }
                    )
                }
                
                
                DESEQ2_SECTION_HEADER (
                    "star_rsem"
                )
                ch_multiqc_files = ch_multiqc_files.mix(DESEQ2_SECTION_HEADER.out.section_header)
                ch_versions = ch_versions.mix(DESEQ2_SECTION_HEADER.out.versions)
                
                
                DESEQ2_TRANSFORM (
                    ch_deseq2_raw_files.flatten(),
                    ch_deseq2_pca_header,
                    ch_deseq2_clustering_header,
                    ch_deseq2_read_dist_header
                )
                ch_multiqc_files = ch_multiqc_files.mix(DESEQ2_TRANSFORM.out.multiqc_files)
                ch_versions = ch_versions.mix(DESEQ2_TRANSFORM.out.versions.first())
                
                
                ch_versions = ch_versions.mix(ch_normalization_versions)
                ch_scaling_factors = ch_scaling_factors.mix(ch_normalization_scaling_factors)
            }

        }
        
        
        if (quantification_methods.contains('genome')) {
            //

            //
            GENOME_COUNT (
                ch_genome_bam.join(ch_genome_bam_index, by: [0]),
                ch_genomic_features
            )
            ch_versions = ch_versions.mix(GENOME_COUNT.out.versions)

            MERGE_GENOME_COUNTS (
                GENOME_COUNT.out.combined_counts.map { meta, counts -> counts }.collect(),
                "transcript,intron,exon,5utr,3utr",
                ch_annotation_matrix
            )
            ch_versions = ch_versions.mix(MERGE_GENOME_COUNTS.out.versions)

            // Generate genome counts text report
            GENOME_COUNTS_REPORT (
                GENOME_COUNT.out.transcript_counts.map { meta, file -> file }
                    .mix(GENOME_COUNT.out.intron_counts.map { meta, file -> file })
                    .mix(GENOME_COUNT.out.exon_counts.map { meta, file -> file })
                    .mix(GENOME_COUNT.out.utr5_counts.map { meta, file -> file })
                    .mix(GENOME_COUNT.out.utr3_counts.map { meta, file -> file })
                    .collect()
            )
            ch_versions = ch_versions.mix(GENOME_COUNTS_REPORT.out.versions)

            if (!params.skip_qc & !params.skip_deseq2_qc) {
                
                def normalization_methods = params.normalization_method instanceof List ? 
                    params.normalization_method : params.normalization_method.split(',').collect{it.trim()}
                
                
                ch_normalization_versions_genome = Channel.empty()
                ch_normalization_scaling_factors_genome = Channel.empty()
                ch_deseq2_raw_files_genome = Channel.empty()
                
                
                if (normalization_methods.contains('invariant_genes')) {
                    NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME (
                        MERGE_GENOME_COUNTS.out.merged_counts.flatten().filter { it.name.contains('genome_exon_counts_merged.txt') }.first(),
                        "STAR_Genome"
                    )
                    
                    
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.read_dist_norm_txt)
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.sample_distances_txt)
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.pca_all_genes_txt)
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.pca_top_genes_txt)
                    
                    ch_normalization_versions_genome = ch_normalization_versions_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.versions)
                    ch_normalization_scaling_factors_genome = ch_normalization_scaling_factors_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.scaling_factors)
                    ch_normalization_scaling_factors_genome = ch_normalization_scaling_factors_genome.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.scaling_factors_individual)
                    ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.scaling_factors_individual)
                    // Tag files with quantification method for downstream use
                    ch_scaling_factors_individual_invariant = ch_scaling_factors_individual_invariant.mix(
                        NORMALIZE_DESEQ2_QC_INVARIANT_GENES_GENOME.out.scaling_factors_individual
                            .flatten()
                            .map { file -> [file, 'genome'] }
                    )
                }
                
                
                if (normalization_methods.contains('all_genes') || (!normalization_methods.contains('invariant_genes') && !normalization_methods.contains('all_genes'))) {
                    NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME (
                        MERGE_GENOME_COUNTS.out.merged_counts.flatten().filter { it.name.contains('genome_exon_counts_merged.txt') }.first(),
                        "STAR_Genome"
                    )
                    
                    
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.read_dist_norm_txt)
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.sample_distances_txt)
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.pca_all_genes_txt)
                    ch_deseq2_raw_files_genome = ch_deseq2_raw_files_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.pca_top_genes_txt)
                    
                    ch_normalization_versions_genome = ch_normalization_versions_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.versions)
                    ch_normalization_scaling_factors_genome = ch_normalization_scaling_factors_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.scaling_factors)
                    ch_normalization_scaling_factors_genome = ch_normalization_scaling_factors_genome.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.scaling_factors_individual)
                    ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.scaling_factors_individual)
                    // Tag files with quantification method for downstream use
                    ch_scaling_factors_individual_all_genes = ch_scaling_factors_individual_all_genes.mix(
                        NORMALIZE_DESEQ2_QC_ALL_GENES_GENOME.out.scaling_factors_individual
                            .flatten()
                            .map { file -> [file, 'genome'] }
                    )
                }
                
                
                DESEQ2_SECTION_HEADER_STAR_GENOME (
                    "star_genome"
                )
                ch_multiqc_files = ch_multiqc_files.mix(DESEQ2_SECTION_HEADER_STAR_GENOME.out.section_header)
                ch_multiqc_star_quant = ch_multiqc_star_quant.mix(DESEQ2_SECTION_HEADER_STAR_GENOME.out.section_header)
                ch_versions = ch_versions.mix(DESEQ2_SECTION_HEADER_STAR_GENOME.out.versions)
                
                
                DESEQ2_TRANSFORM_STAR_GENOME (
                    ch_deseq2_raw_files_genome.flatten(),
                    ch_deseq2_pca_header,
                    ch_deseq2_clustering_header,
                    ch_deseq2_read_dist_header
                )
                ch_multiqc_files = ch_multiqc_files.mix(DESEQ2_TRANSFORM_STAR_GENOME.out.multiqc_files)
                ch_multiqc_star_quant = ch_multiqc_star_quant.mix(DESEQ2_TRANSFORM_STAR_GENOME.out.multiqc_files)
                
                
                ch_versions = ch_versions.mix(ch_normalization_versions_genome)
                ch_versions = ch_versions.mix(DESEQ2_TRANSFORM_STAR_GENOME.out.versions.first())
                ch_scaling_factors = ch_scaling_factors.mix(ch_normalization_scaling_factors_genome)            }
        }
    }

    //
    //
    if (!params.skip_alignment && params.aligner == 'hisat2') {
        FASTQ_ALIGN_HISAT2 (
            ch_strand_inferred_filtered_fastq,
            ch_hisat2_index.map { [ [:], it ] },
            ch_splicesites.map { [ [:], it ] },
            ch_fasta.map { [ [:], it ] }
        )
        // Separate HISAT2 BAM outputs for aligner-specific QC
        ch_genome_bam_hisat2      = ch_genome_bam_hisat2.mix(FASTQ_ALIGN_HISAT2.out.bam)
        ch_genome_bam_hisat2_index = ch_genome_bam_hisat2_index.mix(params.bam_csi_index ? FASTQ_ALIGN_HISAT2.out.csi : FASTQ_ALIGN_HISAT2.out.bai)
        // Keep old channels for backward compatibility
        ch_genome_bam          = ch_genome_bam.mix(FASTQ_ALIGN_HISAT2.out.bam)
        ch_genome_bam_index    = ch_genome_bam_index.mix(params.bam_csi_index ? FASTQ_ALIGN_HISAT2.out.csi : FASTQ_ALIGN_HISAT2.out.bai)
        ch_unprocessed_bams    = ch_genome_bam.map { meta, bam -> [ meta, bam, '' ] }
        ch_unaligned_sequences = FASTQ_ALIGN_HISAT2.out.fastq
        ch_multiqc_hisat2_files = ch_multiqc_hisat2_files.mix(FASTQ_ALIGN_HISAT2.out.summary.collect{it[1]})
        ch_multiqc_hisat2_qc    = ch_multiqc_hisat2_qc.mix(FASTQ_ALIGN_HISAT2.out.summary.collect{it[1]})

        ch_versions = ch_versions.mix(FASTQ_ALIGN_HISAT2.out.versions)

        //
        //

        if (params.with_umi) {

            BAM_DEDUP_UMI_HISAT2(
                ch_genome_bam.join(ch_genome_bam_index, by: [0]),
                ch_fasta.map { [ [:], it ] },
                params.umi_dedup_tool,
                params.umitools_dedup_stats,
                params.bam_csi_index,
                ch_transcriptome_bam,
                ch_transcript_fasta.map { [ [:], it ] }
            )

            ch_genome_bam        = BAM_DEDUP_UMI_HISAT2.out.bam
            ch_genome_bam_index  = params.bam_csi_index ? BAM_DEDUP_UMI_HISAT2.out.csi : BAM_DEDUP_UMI_HISAT2.out.bai
            ch_versions          = ch_versions.mix(BAM_DEDUP_UMI_HISAT2.out.versions)

            ch_multiqc_files = ch_multiqc_files
                .mix(BAM_DEDUP_UMI_HISAT2.out.multiqc_files)
        } else {

            ch_multiqc_files = ch_multiqc_files
                .mix(FASTQ_ALIGN_HISAT2.out.stats.collect{it[1]})
                .mix(FASTQ_ALIGN_HISAT2.out.flagstat.collect{it[1]})
                .mix(FASTQ_ALIGN_HISAT2.out.idxstats.collect{it[1]})
        }

        //
        //
        if (quantification_methods.contains('genome') && !params.skip_quantification_method) {
            GENOME_COUNT (
                ch_genome_bam.join(ch_genome_bam_index, by: [0]),
                ch_genomic_features
            )
            ch_versions = ch_versions.mix(GENOME_COUNT.out.versions)

            MERGE_GENOME_COUNTS_HISAT2 (
                GENOME_COUNT.out.combined_counts.map { meta, counts -> counts }.collect(),
                "transcript,intron,exon,5utr,3utr",
                ch_annotation_matrix
            )
            ch_versions = ch_versions.mix(MERGE_GENOME_COUNTS_HISAT2.out.versions)

            // Generate genome counts text report
            GENOME_COUNTS_REPORT (
                GENOME_COUNT.out.transcript_counts.map { meta, file -> file }
                    .mix(GENOME_COUNT.out.intron_counts.map { meta, file -> file })
                    .mix(GENOME_COUNT.out.exon_counts.map { meta, file -> file })
                    .mix(GENOME_COUNT.out.utr5_counts.map { meta, file -> file })
                    .mix(GENOME_COUNT.out.utr3_counts.map { meta, file -> file })
                    .collect()
            )
            ch_versions = ch_versions.mix(GENOME_COUNTS_REPORT.out.versions)

            if (!params.skip_qc & !params.skip_deseq2_qc) {
                
                def normalization_methods = params.normalization_method instanceof List ? 
                    params.normalization_method : params.normalization_method.split(',').collect{it.trim()}
                
                
                ch_normalization_versions_hisat2 = Channel.empty()
                ch_normalization_scaling_factors_hisat2 = Channel.empty()
                ch_deseq2_raw_files_hisat2 = Channel.empty()
                
                
                if (normalization_methods.contains('invariant_genes')) {
                    NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT (
                        MERGE_GENOME_COUNTS_HISAT2.out.merged_counts.flatten().filter { it.name.contains('genome_exon_counts_merged.txt') }.first(),
                        "HISAT2_Genome"
                    )
                    
                    
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.read_dist_norm_txt)
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.sample_distances_txt)
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.pca_all_genes_txt)
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.pca_top_genes_txt)
                    
                    ch_normalization_versions_hisat2 = ch_normalization_versions_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.versions)
                    ch_normalization_scaling_factors_hisat2 = ch_normalization_scaling_factors_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.scaling_factors)
                    ch_normalization_scaling_factors_hisat2 = ch_normalization_scaling_factors_hisat2.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.scaling_factors_individual)
                    ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.scaling_factors_individual)
                    // Tag files with quantification method for downstream use
                    ch_scaling_factors_individual_invariant = ch_scaling_factors_individual_invariant.mix(
                        NORMALIZE_DESEQ2_QC_INVARIANT_GENES_ALIGNMENT.out.scaling_factors_individual
                            .flatten()
                            .map { file -> [file, 'genome'] }
                    )
                }
                
                
                if (normalization_methods.contains('all_genes') || (!normalization_methods.contains('invariant_genes') && !normalization_methods.contains('all_genes'))) {
                    NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT (
                        MERGE_GENOME_COUNTS_HISAT2.out.merged_counts.flatten().filter { it.name.contains('genome_exon_counts_merged.txt') }.first(),
                        "HISAT2_Genome"
                    )
                    
                    
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.read_dist_norm_txt)
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.sample_distances_txt)
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.pca_all_genes_txt)
                    ch_deseq2_raw_files_hisat2 = ch_deseq2_raw_files_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.pca_top_genes_txt)
                    
                    ch_normalization_versions_hisat2 = ch_normalization_versions_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.versions)
                    ch_normalization_scaling_factors_hisat2 = ch_normalization_scaling_factors_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.scaling_factors)
                    ch_normalization_scaling_factors_hisat2 = ch_normalization_scaling_factors_hisat2.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.scaling_factors_individual)
                    ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.scaling_factors_individual)
                    // Tag files with quantification method for downstream use
                    ch_scaling_factors_individual_all_genes = ch_scaling_factors_individual_all_genes.mix(
                        NORMALIZE_DESEQ2_QC_ALL_GENES_ALIGNMENT.out.scaling_factors_individual
                            .flatten()
                            .map { file -> [file, 'genome'] }
                    )
                }
                
                
                DESEQ2_SECTION_HEADER_HISAT2 (
                    "hisat2_genome"
                )
                ch_multiqc_files = ch_multiqc_files.mix(DESEQ2_SECTION_HEADER_HISAT2.out.section_header)
                ch_multiqc_hisat2_quant = ch_multiqc_hisat2_quant.mix(DESEQ2_SECTION_HEADER_HISAT2.out.section_header)
                ch_versions = ch_versions.mix(DESEQ2_SECTION_HEADER_HISAT2.out.versions)
                
                
                DESEQ2_TRANSFORM_HISAT2 (
                    ch_deseq2_raw_files_hisat2.flatten(),
                    ch_deseq2_pca_header,
                    ch_deseq2_clustering_header,
                    ch_deseq2_read_dist_header
                )
                ch_multiqc_files = ch_multiqc_files.mix(DESEQ2_TRANSFORM_HISAT2.out.multiqc_files)
                ch_multiqc_hisat2_quant = ch_multiqc_hisat2_quant.mix(DESEQ2_TRANSFORM_HISAT2.out.multiqc_files)
                ch_versions = ch_versions.mix(DESEQ2_TRANSFORM_HISAT2.out.versions.first())
                
                
                ch_versions = ch_versions.mix(ch_normalization_versions_hisat2)
                ch_scaling_factors = ch_scaling_factors.mix(ch_normalization_scaling_factors_hisat2)            }
        }
    }


    ch_genome_bam_bai_mapping = ch_genome_bam
        .join(ch_genome_bam_index)
        .join(ch_percent_mapped, remainder: true)
        .map{ row ->
            def (meta, bam, index) = row[0..2]
            def percent_mapped = row.size() == 4 ? row[3] : null
            def pass = percent_mapped != null ? percent_mapped >= params.min_mapped_reads.toFloat() : null
            return [ meta, bam, index, percent_mapped, pass ]
        }
        .multiMap { meta, bam, index, percent_mapped, pass ->
            bam: [ meta, bam, index, pass ]
            percent_mapped: [ meta.id, percent_mapped ]
            percent_mapped_pass: [ meta.id, percent_mapped, pass ]
            status: [ meta.id, pass ]
        }

    ch_percent_mapped = ch_genome_bam_bai_mapping.percent_mapped


    ch_map_status = ch_genome_bam_bai_mapping.status
        .filter { id, pass -> pass != null }

    ch_fail_mapping_multiqc = ch_genome_bam_bai_mapping.percent_mapped_pass
        .filter { id, percent_mapped, pass -> pass != null && !pass }
        .map { id, percent_mapped, pass -> [ "${id}\t${percent_mapped}" ] }
        .collect()
        .map {
            tsv_data ->
                def header = ["Sample", "STAR uniquely mapped reads (%)"]
                sample_status_header_multiqc.text + multiqcTsvFromList(tsv_data, header)
        }
        .collectFile(name: 'fail_mapped_samples_mqc.tsv')

    ch_multiqc_files = ch_multiqc_files.mix(ch_fail_mapping_multiqc)


    map_filtered_genome_bam_bai = ch_genome_bam_bai_mapping.bam
        .filter { meta, bam, index, pass -> pass || pass == null }
        .multiMap { meta, bam, index, pass ->
            bam: [ meta, bam ]
            index: [ meta, index ]
        }

    ch_genome_bam = map_filtered_genome_bam_bai.bam
    ch_genome_bam_index = map_filtered_genome_bam_bai.index

    //
    //

    //
    //
    if (!params.skip_qc && !params.skip_preseq) {
        PRESEQ_LCEXTRAP (
            ch_genome_bam
        )
        ch_multiqc_files = ch_multiqc_files.mix(PRESEQ_LCEXTRAP.out.lc_extrap.collect{it[1]})
        ch_versions = ch_versions.mix(PRESEQ_LCEXTRAP.out.versions.first())
    }

    //
    //
    if (!params.skip_markduplicates && !params.with_umi) {
        BAM_MARKDUPLICATES_PICARD (
            ch_genome_bam,
            ch_fasta.map { [ [:], it ] },
            ch_fai.map { [ [:], it ] }
        )
        ch_genome_bam       = BAM_MARKDUPLICATES_PICARD.out.bam
        ch_genome_bam_index = params.bam_csi_index ? BAM_MARKDUPLICATES_PICARD.out.csi : BAM_MARKDUPLICATES_PICARD.out.bai
        ch_multiqc_files = ch_multiqc_files.mix(BAM_MARKDUPLICATES_PICARD.out.stats.collect{it[1]})
        ch_multiqc_files = ch_multiqc_files.mix(BAM_MARKDUPLICATES_PICARD.out.flagstat.collect{it[1]})
        ch_multiqc_files = ch_multiqc_files.mix(BAM_MARKDUPLICATES_PICARD.out.idxstats.collect{it[1]})
        ch_multiqc_files = ch_multiqc_files.mix(BAM_MARKDUPLICATES_PICARD.out.metrics.collect{it[1]})

        ch_versions = ch_versions.mix(BAM_MARKDUPLICATES_PICARD.out.versions)
    }

    //
    //
    if (!params.skip_stringtie) {
        STRINGTIE_STRINGTIE (
            ch_genome_bam,
            ch_gtf
        )
        ch_versions = ch_versions.mix(STRINGTIE_STRINGTIE.out.versions.first())
    }

    //
    //
    def biotype = params.gencode ? "gene_type" : params.featurecounts_group_type
    if (!params.skip_qc && !params.skip_biotype_qc && biotype) {

        ch_gtf
            .map { biotypeInGtf(it, biotype) }
            .set { biotype_in_gtf }

        ch_genome_bam
            .combine(ch_gtf)
            .combine(biotype_in_gtf)
            .filter { it[-1] }
            .map { it[0..<it.size()-1] }
            .set { ch_featurecounts }

        SUBREAD_FEATURECOUNTS_EXTRA (
            ch_featurecounts
        )
        ch_versions = ch_versions.mix(SUBREAD_FEATURECOUNTS_EXTRA.out.versions.first())

        MULTIQC_CUSTOM_BIOTYPE (
            SUBREAD_FEATURECOUNTS_EXTRA.out.counts,
            ch_biotypes_header_multiqc
        )
        ch_multiqc_files = ch_multiqc_files.mix(MULTIQC_CUSTOM_BIOTYPE.out.tsv.collect{it[1]})
        ch_versions = ch_versions.mix(MULTIQC_CUSTOM_BIOTYPE.out.versions.first())
    }

    //
    //
    if (!params.skip_bigwig) {

        ch_genomecov_input = ch_genome_bam.map { meta, bam -> [ meta, bam, 1 ] }

        BEDTOOLS_GENOMECOV_FW (
            ch_genomecov_input,
            [],
            'bedGraph',
            true
        )
        BEDTOOLS_GENOMECOV_REV (
            ch_genomecov_input,
            [],
            'bedGraph',
            true
        )
        BEDTOOLS_GENOMECOV_UNSTRANDED (
            ch_genomecov_input,
            [],
            'bedGraph',
            true
        )

        ch_versions = ch_versions.mix(BEDTOOLS_GENOMECOV_FW.out.versions.first())

        //
        //
        BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_FORWARD (
            BEDTOOLS_GENOMECOV_FW.out.genomecov,
            ch_chrom_sizes
        )
        ch_versions = ch_versions.mix(BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_FORWARD.out.versions)

        BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_REVERSE (
            BEDTOOLS_GENOMECOV_REV.out.genomecov,
            ch_chrom_sizes
        )

        BEDGRAPH_BEDCLIP_BEDGRAPHTOBIGWIG_UNSTRANDED (
            BEDTOOLS_GENOMECOV_UNSTRANDED.out.genomecov,
            ch_chrom_sizes
        )

    }


    //
    //
    if (!params.skip_qc) {
        if (!params.skip_qualimap) {
            QUALIMAP_RNASEQ (
                ch_genome_bam,
                ch_gtf.map { [ [:], it ] }
            )
            ch_multiqc_files = ch_multiqc_files.mix(QUALIMAP_RNASEQ.out.results.collect{it[1]})
            ch_versions = ch_versions.mix(QUALIMAP_RNASEQ.out.versions.first())
        }

        if (!params.skip_dupradar) {
            DUPRADAR (
                ch_genome_bam,
                ch_gtf.map { [ [:], it ] }
            )
            ch_multiqc_files = ch_multiqc_files.mix(DUPRADAR.out.multiqc.collect{it[1]})
            ch_versions = ch_versions.mix(DUPRADAR.out.versions.first())
        }

        def rseqc_modules = params.rseqc_modules ? params.rseqc_modules.split(',').collect{ it.trim().toLowerCase() } : []
        if (params.bam_csi_index) {
            for (rseqc_module in ['read_distribution', 'inner_distance', 'tin']) {
                if (rseqc_modules.contains(rseqc_module)) {
                    rseqc_modules.remove(rseqc_module)
                }
            }
        }
        if (!params.skip_rseqc && rseqc_modules.size() > 0) {
            BAM_RSEQC (
                ch_genome_bam.join(ch_genome_bam_index, by: [0]),
                ch_gene_bed,
                rseqc_modules
            )
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.bamstat_txt.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.inferexperiment_txt.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.innerdistance_freq.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.junctionannotation_log.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.junctionsaturation_rscript.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.readdistribution_txt.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.readduplication_pos_xls.collect{it[1]})
            ch_multiqc_files = ch_multiqc_files.mix(BAM_RSEQC.out.tin_txt.collect{it[1]})
            ch_versions = ch_versions.mix(BAM_RSEQC.out.versions)

            ch_strand_comparison = BAM_RSEQC.out.inferexperiment_txt
                .map {
                    meta, strand_log ->
                        def rseqc_inferred_strand = getInferexperimentStrandedness(strand_log, params.stranded_threshold, params.unstranded_threshold)
                        rseqc_strandedness = rseqc_inferred_strand.inferred_strandedness

                        def status = 'fail'
                        def multiqc_lines = []
                        if (meta.salmon_strand_analysis) {
                            salmon_strandedness = meta.salmon_strand_analysis.inferred_strandedness

                            if (salmon_strandedness == rseqc_strandedness && rseqc_strandedness != 'undetermined') {
                                status = 'pass'
                            }
                            multiqc_lines = [
                                "$meta.id \tSalmon\t$status\tauto\t${meta.salmon_strand_analysis.values().join('\t')}",
                                "$meta.id\tRSeQC\t$status\tauto\t${rseqc_inferred_strand.values().join('\t')}"
                            ]
                        }
                        else {
                            if (meta.strandedness == rseqc_strandedness) {
                                status = 'pass'
                            }
                            multiqc_lines = [ "$meta.id\tRSeQC\t$status\t$meta.strandedness\t${rseqc_inferred_strand.values().join('\t')}" ]
                        }
                        return [ meta, status, multiqc_lines ]
                }
                .multiMap {
                    meta, status, multiqc_lines ->
                        status: [ meta.id, status == 'pass' ]
                        multiqc_lines: multiqc_lines
                }

            ch_strand_status = ch_strand_comparison.status

            ch_strand_comparison.multiqc_lines
                .flatten()
                .collect()
                .map {
                    tsv_data ->
                        def header = [
                            "Sample",
                            "Strand inference method",
                            "Status",
                            "Provided strandedness",
                            "Inferred strandedness",
                            "Sense (%)",
                            "Antisense (%)",
                            "Unstranded (%)"
                        ]
                        sample_status_header_multiqc.text + multiqcTsvFromList(tsv_data, header)
                }
                .set { ch_fail_strand_multiqc }

            ch_multiqc_files = ch_multiqc_files.mix(ch_fail_strand_multiqc.collectFile(name: 'fail_strand_check_mqc.tsv'))
        }

        if (params.contaminant_screening in ['kraken2', 'kraken2_bracken'] ) {
            KRAKEN2 (
                ch_unaligned_sequences,
                params.kraken_db,
                params.save_kraken_assignments,
                params.save_kraken_unassigned
            )
            ch_kraken_reports = KRAKEN2.out.report
            ch_versions = ch_versions.mix(KRAKEN2.out.versions)

            if (params.contaminant_screening == 'kraken2') {
                ch_multiqc_files = ch_multiqc_files.mix(KRAKEN2.out.report.collect{it[1]})
            } else if (params.contaminant_screening == 'kraken2_bracken') {
                BRACKEN (
                    ch_kraken_reports,
                    params.kraken_db
                )
                ch_versions = ch_versions.mix(BRACKEN.out.versions)
                ch_multiqc_files = ch_multiqc_files.mix(BRACKEN.out.txt.collect{it[1]})
            }
        }
    }

    //
    //
    if (!params.skip_pseudo_alignment && params.pseudo_aligner) {

        // Only kallisto is supported as pseudo-aligner
        ch_pseudo_index = ch_kallisto_index

        QUANTIFY_PSEUDO_ALIGNMENT (
            ch_samplesheet.map { [ [:], it ] },
            ch_strand_inferred_filtered_fastq,
            ch_pseudo_index,
            ch_dummy_file,
            ch_gtf,
            params.gtf_group_features,
            params.gtf_extra_attributes,
            params.pseudo_aligner,
            false,
            '',
            params.kallisto_quant_fraglen,
            params.kallisto_quant_fraglen_sd,
            ch_annotation_matrix,
            ch_chrom_sizes
        )
        ch_counts_gene = QUANTIFY_PSEUDO_ALIGNMENT.out.counts_gene
        ch_versions = ch_versions.mix(QUANTIFY_PSEUDO_ALIGNMENT.out.versions)
        
        // Only Kallisto is a true pseudo-aligner (salmon used with STAR is quantification, not pseudo-alignment)
        if (params.pseudo_aligner == 'kallisto') {
            ch_multiqc_kallisto_files = ch_multiqc_kallisto_files.mix(QUANTIFY_PSEUDO_ALIGNMENT.out.multiqc.collect{it[1]})
            ch_genome_bam = ch_genome_bam.mix(QUANTIFY_PSEUDO_ALIGNMENT.out.bam)
            ch_genome_bam_index = ch_genome_bam_index.mix(QUANTIFY_PSEUDO_ALIGNMENT.out.bai)
        }

        //
        //
        if (!params.skip_qc & !params.skip_deseq2_qc) {
            
            def normalization_methods = params.normalization_method instanceof List ? 
                params.normalization_method : params.normalization_method.split(',').collect{it.trim()}
            
            
            ch_normalization_versions_pseudo = Channel.empty()
            ch_normalization_scaling_factors_pseudo = Channel.empty()
            ch_deseq2_raw_files_pseudo = Channel.empty()
            
            def pseudo_quantifier = params.pseudo_aligner ? params.pseudo_aligner.capitalize() : "Salmon"
            
            
            if (normalization_methods.contains('invariant_genes')) {
                NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO (
                    ch_counts_gene,
                    pseudo_quantifier
                )

                
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.read_dist_norm_txt)
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.sample_distances_txt)
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.pca_all_genes_txt)
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.pca_top_genes_txt)
                
                ch_normalization_versions_pseudo = ch_normalization_versions_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.versions)
                ch_normalization_scaling_factors_pseudo = ch_normalization_scaling_factors_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.scaling_factors)
                ch_normalization_scaling_factors_pseudo = ch_normalization_scaling_factors_pseudo.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.scaling_factors_individual)
                ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.scaling_factors_individual)
                // Tag files with quantification method for downstream use (salmon or kallisto)
                ch_scaling_factors_individual_invariant = ch_scaling_factors_individual_invariant.mix(
                    NORMALIZE_DESEQ2_QC_INVARIANT_GENES_PSEUDO.out.scaling_factors_individual
                        .flatten()
                        .map { file -> [file, pseudo_quantifier.toLowerCase()] }
                )
            }
            
            
            if (normalization_methods.contains('all_genes') || (!normalization_methods.contains('invariant_genes') && !normalization_methods.contains('all_genes'))) {
                NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO (
                    ch_counts_gene,
                    pseudo_quantifier
                )

                
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.read_dist_norm_txt)
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.sample_distances_txt)
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.pca_all_genes_txt)
                ch_deseq2_raw_files_pseudo = ch_deseq2_raw_files_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.pca_top_genes_txt)
                
                ch_normalization_versions_pseudo = ch_normalization_versions_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.versions)
                ch_normalization_scaling_factors_pseudo = ch_normalization_scaling_factors_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.scaling_factors)
                ch_normalization_scaling_factors_pseudo = ch_normalization_scaling_factors_pseudo.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.scaling_factors_individual)
                ch_scaling_factors_individual = ch_scaling_factors_individual.mix(NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.scaling_factors_individual)
                // Tag files with quantification method for downstream use (salmon or kallisto)
                ch_scaling_factors_individual_all_genes = ch_scaling_factors_individual_all_genes.mix(
                    NORMALIZE_DESEQ2_QC_ALL_GENES_PSEUDO.out.scaling_factors_individual
                        .flatten()
                        .map { file -> [file, pseudo_quantifier.toLowerCase()] }
                )
            }
            
            
            DESEQ2_SECTION_HEADER_PSEUDO (
                params.pseudo_aligner
            )
            ch_multiqc_kallisto_files = ch_multiqc_kallisto_files.mix(DESEQ2_SECTION_HEADER_PSEUDO.out.section_header)
            ch_versions = ch_versions.mix(DESEQ2_SECTION_HEADER_PSEUDO.out.versions)
            
            
            DESEQ2_TRANSFORM_PSEUDO (
                ch_deseq2_raw_files_pseudo.flatten(),
                ch_deseq2_pca_header,
                ch_deseq2_clustering_header,
                ch_deseq2_read_dist_header
            )
            ch_multiqc_kallisto_files = ch_multiqc_kallisto_files.mix(DESEQ2_TRANSFORM_PSEUDO.out.multiqc_files)
            
            ch_versions = ch_versions.mix(DESEQ2_TRANSFORM_PSEUDO.out.versions.first())
            ch_versions = ch_versions.mix(ch_normalization_versions_pseudo)
            ch_scaling_factors = ch_scaling_factors.mix(ch_normalization_scaling_factors_pseudo)        }
    }

    //
    // MODULE: Generate normalized BigWig files using DESeq2 scaling factors
    // Uses mmrnaseq strategy: read scaling factors once, combine with BAM files, apply normalization
    //
    if (!params.skip_deeptools_norm && !params.skip_qc && !params.skip_deseq2_qc) {
        def normalization_methods = params.normalization_method instanceof List ? 
            params.normalization_method : params.normalization_method.split(',').collect{it.trim()}
        
        // Prepare BAM channel with BAI index files
        ch_bam_for_deeptools = ch_genome_bam
            .join(ch_genome_bam_index, by: [0])
        
        if (normalization_methods.contains('invariant_genes')) {
            //
            // CHANNEL OPERATION: Extract scaling factors from invariant genes normalization
            // Use dedicated invariant genes channel instead of filtering mixed channel
            // Extract sample name, scaling value, and quantification method as tuple: [sample_name, scaling_value, quant_method]
            //
            ch_scaling_per_sample_invariant = ch_scaling_factors_individual_invariant
                .map { file, quant_method ->
                    // Channel already contains [file, quant_method] tuples from source
                    def sample_name = file.name.replaceAll('_scaling_factor\\.txt$', '')
                    def scaling_value = file.text.trim()
                    
                    [sample_name, scaling_value, quant_method]
                }
            
            //
            // CHANNEL OPERATION: Combine BAM files with scaling factors (mmrnaseq strategy)
            // Use .combine() to create cartesian product, then filter by matching sample IDs
            // Result: [meta, bam, bai, scaling_value]
            // Add quantification method to meta map for publishDir routing
            //
            ch_combined_input_invariant = ch_bam_for_deeptools
                .combine(ch_scaling_per_sample_invariant)
                .map { meta, bam, bai, sample_id, scaling, quant_method -> 
                    if (meta.id == sample_id) {
                        def new_meta = meta.clone()
                        new_meta.quantification = quant_method
                        [new_meta, bam, bai, scaling]
                    } else {
                        null
                    }
                }
                .filter { it != null }
            
            // Debug output for invariant genes channel
            ch_combined_input_invariant
                .view { meta, bam, bai, scaling -> 
                    "DEEPTOOLS_INVARIANT: Sample=${meta.id}, Quant=${meta.quantification}, Scaling=${scaling}, BAM=${bam.name}" 
                }
                .set { ch_combined_input_invariant_debug }
            
            DEEPTOOLS_BIGWIG_NORM_INVARIANT (
                ch_combined_input_invariant_debug
            )
            ch_versions = ch_versions.mix(DEEPTOOLS_BIGWIG_NORM_INVARIANT.out.versions)
        }
        
        if (normalization_methods.contains('all_genes') || (!normalization_methods.contains('invariant_genes') && !normalization_methods.contains('all_genes'))) {
            //
            // CHANNEL OPERATION: Extract scaling factors from all genes normalization
            // Use dedicated all genes channel instead of filtering mixed channel
            // Extract sample name, scaling value, and quantification method as tuple: [sample_name, scaling_value, quant_method]
            //
            ch_scaling_per_sample_all_genes = ch_scaling_factors_individual_all_genes
                .map { file, quant_method ->
                    // Channel already contains [file, quant_method] tuples from source
                    def sample_name = file.name.replaceAll('_scaling_factor\\.txt$', '')
                    def scaling_value = file.text.trim()
                    
                    [sample_name, scaling_value, quant_method]
                }
            
            //
            // CHANNEL OPERATION: Combine BAM files with scaling factors (mmrnaseq strategy)
            // Use .combine() to create cartesian product, then filter by matching sample IDs
            // Result: [meta, bam, bai, scaling_value]
            // Add quantification method to meta map for publishDir routing
            //
            ch_combined_input_all_genes = ch_bam_for_deeptools
                .combine(ch_scaling_per_sample_all_genes)
                .map { meta, bam, bai, sample_id, scaling, quant_method -> 
                    if (meta.id == sample_id) {
                        def new_meta = meta.clone()
                        new_meta.quantification = quant_method
                        [new_meta, bam, bai, scaling]
                    } else {
                        null
                    }
                }
                .filter { it != null }
            
            // Debug output for all genes channel
            ch_combined_input_all_genes
                .view { meta, bam, bai, scaling -> 
                    "DEEPTOOLS_ALL_GENES: Sample=${meta.id}, Quant=${meta.quantification}, Scaling=${scaling}, BAM=${bam.name}" 
                }
                .set { ch_combined_input_all_genes_debug }
            
            DEEPTOOLS_BIGWIG_NORM_ALL_GENES (
                ch_combined_input_all_genes_debug
            )
            ch_versions = ch_versions.mix(DEEPTOOLS_BIGWIG_NORM_ALL_GENES.out.versions)
        }
    }

    //
    // MODULE: Collate software versions from all tools used in the pipeline
    //
    softwareVersionsToYAML(ch_versions)
        .collectFile(storeDir: "${params.outdir}/pipeline_info", name: 'nf_core_rnaseq_software_mqc_versions.yml', sort: true, newLine: true)
        .set { ch_collated_versions }

    //
    // MODULE: MultiQC - Aggregate all QC reports into a single HTML report
    //
    ch_multiqc_report = Channel.empty()

    if (!params.skip_multiqc) {

        ch_multiqc_config        = Channel.fromPath("$projectDir/workflows/rnaseq/assets/multiqc/multiqc_config.yml", checkIfExists: true)
        ch_multiqc_custom_config = params.multiqc_config ? Channel.fromPath(params.multiqc_config) : Channel.empty()
        ch_multiqc_logo          = params.multiqc_logo   ? Channel.fromPath(params.multiqc_logo)   : Channel.empty()

        ch_workflow_summary = Channel.value(
            paramsSummaryMultiqc(
                paramsSummaryMap(workflow, parameters_schema: "nextflow_schema.json")
            )
        ).collectFile(name: 'workflow_summary_mqc.yaml')

        ch_methods_description = Channel.value(
            methodsDescriptionText(
                params.multiqc_methods_description
                    ? file(params.multiqc_methods_description)
                    : file("$projectDir/workflows/rnaseq/assets/multiqc/methods_description_template.yml", checkIfExists: true)
            )
        ).collectFile(name: 'methods_description_mqc.yaml')

        ch_multiqc_files = ch_multiqc_files
            .mix(ch_workflow_summary)
            .mix(ch_collated_versions)
            .mix(ch_methods_description)


        ch_name_replacements = ch_fastq
            .map{ meta, reads ->
                def name1 = file(reads[0][0]).simpleName + "\t" + meta.id + '_1'
                def fastqcnames = meta.id + "_raw\t" + meta.id + "\n" + meta.id + "_trimmed\t" + meta.id
                if (reads[0][1] ){
                    def name2 = file(reads[0][1]).simpleName + "\t" + meta.id + '_2'
                    def fastqcnames1 = meta.id + "_raw_1\t" + meta.id + "_1\n" + meta.id + "_trimmed_1\t" + meta.id + "_1"
                    def fastqcnames2 = meta.id + "_raw_2\t" + meta.id + "_2\n" + meta.id + "_trimmed_2\t" + meta.id + "_2"
                    return [ name1, name2, fastqcnames1, fastqcnames2 ]
                } else{
                    return [ name1, fastqcnames ]
                }
            }
            .flatten()
            .collectFile(name: 'name_replacement.txt', newLine: true)
            .ifEmpty([])

        // Generate comprehensive MultiQC report combining all QC files
        // Collects files from all aligners and tools used in the workflow
        MULTIQC (
            ch_multiqc_files.mix(ch_multiqc_star_files).mix(ch_multiqc_hisat2_files).mix(ch_multiqc_kallisto_files).mix(ch_multiqc_star_quant).mix(ch_multiqc_hisat2_quant).collect(),
            ch_multiqc_config.toList(),
            ch_multiqc_custom_config.toList(),
            ch_multiqc_logo.toList(),
            ch_name_replacements,
            []
        )
        ch_multiqc_report = MULTIQC.out.report

    }

    //
    //

    ch_samplesheet_with_bams = Channel.empty()
    if (!params.skip_alignment && params.save_align_intermeds) {
        ch_fastq.map { meta, reads -> [ meta.id, meta, reads ] }
            .join(ch_unprocessed_bams.map { meta, genome_bam, transcriptome_bam -> [ meta.id, meta, genome_bam, transcriptome_bam ] })
            .join(ch_percent_mapped)
            .transpose()
            .map { id, fastq_meta, reads, meta, genome_bam, transcriptome_bam, percent_mapped ->

                def genome_bam_published = meta.has_genome_bam ?
                    (meta.original_genome_bam ?: '') :
                    mapBamToPublishedPath(genome_bam, meta.id, params.aligner, params.outdir)

                def transcriptome_bam_published = meta.has_transcriptome_bam ?
                    (meta.original_transcriptome_bam ?: '') :
                    mapBamToPublishedPath(transcriptome_bam, meta.id, params.aligner, params.outdir)

                def fastq_1 = reads[0].toUriString()
                def fastq_2 = reads.size() > 1 ? reads[1].toUriString() : ''
                def mapped = percent_mapped != null ? percent_mapped : ''

                return "${meta.id},${fastq_1},${fastq_2},${meta.strandedness},${genome_bam_published},${mapped},${transcriptome_bam_published}"
            }
            .collectFile(
                name: 'samplesheet_with_bams.csv',
                storeDir: "${params.outdir}/samplesheets",
                newLine: true,
                seed: 'sample,fastq_1,fastq_2,strandedness,genome_bam,percent_mapped,transcriptome_bam'
            )
    }

    emit:
    trim_status    = ch_trim_status    // channel: [id, boolean]
    map_status     = ch_map_status     // channel: [id, boolean]
    strand_status  = ch_strand_status  // channel: [id, boolean]
    multiqc_report = ch_multiqc_report // channel: /path/to/multiqc_report.html
    versions       = ch_versions       // channel: [ path(versions.yml) ]
}

/*
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    THE END
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
*/
